### PR TITLE
Short-circuit painless def equality

### PR DESCRIPTION
**Pull request by thecoop**
Add short-circuiting checks to `Def.eq` implementation.

This shows a 8% increase in throughput in the index-append-with-ingest-baseline-pipeline rally test, 15% improvement in security/logs track

### CODE REVIEW

**Comment by thecoop** [comment]
@elasticsearchmachine rerun elasticsearch-ci/packaging-tests-windows-sample

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java [hunk 1]
```diff
@@ -549,7 +553,7 @@ private static boolean eq(Object left, Object right) {
                 if (left instanceof Number) {
                     return ((Number) left).floatValue() == (float) right;
                 } else if (left instanceof Character) {
-                    return (char) left == ((Number) right).floatValue();
+                    return (char) left == (float) right;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java:556]: This is not covered by a test, can you add some coverage?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java [hunk 2]
```diff
@@ -561,7 +565,7 @@ private static boolean eq(Object left, Object right) {
                 if (left instanceof Number) {
                     return ((Number) left).longValue() == (long) right;
                 } else if (left instanceof Character) {
-                    return (char) left == ((Number) right).longValue();
+                    return (char) left == (long) right;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java:568]: We're also missing a test that hits this code, can you add one?

Comment (thecoop) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java:568]: Added

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java (hunk 1 above)
Comment (thecoop) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java:556]: Added

**Comment by thecoop** [comment]
@elasticsearchmachine update branch

**Review by stu-elastic — APPROVED** [review]
lgtm

**Pull request merged**